### PR TITLE
SMHE-1609: Improve Mx382AlarmDecoder handling

### PR DESCRIPTION
**Pull request by abolder**
Accept Mx382AlarmDecoder message with date and time within the message so alarm will be forwarded.

### CODE REVIEW

**Comment by abolder** [comment]
https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/72/

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java [hunk 1]
```diff
@@ -17,72 +17,104 @@ public class Mx382AlarmDecoder extends AlarmDecoder {
...
       throws UnrecognizedMessageDataException {
 
-    this.checkBytes(inputStream, WPDU_HEADER, "WPDU-header");
-    this.checkByte(inputStream, EVENT_NOTIFICATION_REQUEST, "event-notification-request");
+    this.handleWpduHeaderBytes(inputStream, "WPDU-header");
+    this.handleByte(inputStream, EVENT_NOTIFICATION_REQUEST, "event-notification-request");
 
-    // Datetime is not used, so can be skipped as well
-    this.skipDateTime(inputStream);
+    // Datetime is read and copied
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java]: The datetime bytes are read but not used

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java [hunk 2]
```diff
@@ -17,72 +17,104 @@ public class Mx382AlarmDecoder extends AlarmDecoder {
...
+      throws UnrecognizedMessageDataException {
+    // The first byte tells if datetime is provided 0x00 is not provided and 0x09 is provided
+    final byte dataType = this.readByte(inputStream);
+
+    // read the rest of the datetime bytes date is provided
+    if (dataType == DATA_TYPE_OCTET_STRING) {
+      // The next byte of the datetime is the length byte
+      final byte lengthOfData = this.readByte(inputStream);
+      this.readBytes(inputStream, lengthOfData);
+    }
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java:63]: Why are the dateTime bytes not appended?

**Review by harrymiddelburgetall — COMMENTED** [review]
In case of an exception in decoding the message it would be nice to see the whole message instead of only the byte(s) that were not expected.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java [hunk 3]
```diff
@@ -17,72 +17,104 @@ public class Mx382AlarmDecoder extends AlarmDecoder {
...
+
   private void addAlarm() {
     this.builder.withTriggerType(PUSH_ALARM_TRIGGER);
     final Set<AlarmTypeDto> alarmSet = new HashSet<>();
     alarmSet.add(AlarmTypeDto.LAST_GASP);
     this.builder.addAlarms(alarmSet);
   }
 
-  private void checkByte(final InputStream inputStream, final byte expectedByte, final String name)
+  private void handleByte(final InputStream inputStream, final byte expectedByte, final String name)
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java]: Why is this renamed to handleByte? It seems to do the same as checkBytes, but then only for 1 byte.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoderTest.java [hunk 4]
```diff
@@ -39,16 +45,52 @@ void testDecode() throws UnrecognizedMessageDataException {
     assertThat(dlmsPushNotification.getEquipmentIdentifier()).isEqualTo(this.equipmentIdentifier);
   }
 
+  /*
+   * Decode with an alarm message that contains a date time stamp, this has an alternative length withing the WDPU header (2a on index 7, 8th byte).
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoderTest.java]: `within` instead of `withing`

**Review by stefanermens — CHANGES_REQUESTED** [review]

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java (hunk 1 above)
Comment (abolder) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java]: Modified the comment that the datetime is skipped.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java (hunk 2 above)
Comment (abolder) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java:63]: The datetime is not used in this message and for other messages the dates are also skipped.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoderTest.java (hunk 4 above)
Comment (abolder) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoderTest.java]: modified

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java (hunk 3 above)
Comment (abolder) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java]: renamed back to checkBytes

**Review by stefanermens — APPROVED** [review]

**Review by harrymiddelburgetall — APPROVED** [review]

**Comment by abolder** [comment]
https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/73/
